File: ios/GhostChat/Core/Managers/SettingsManager.swift
import Foundation

/// Keychain-backed user preferences. No UserDefaults anywhere.
@MainActor
final class SettingsManager: ObservableObject {

    private struct Keys {
        static let privacyMode          = "settings.privacy_mode"
        static let biometricEnabled     = "settings.biometric_enabled"
        static let soundEnabled         = "settings.sound_enabled"
        static let messageTTL           = "settings.message_ttl"
        static let autoLock             = "settings.auto_lock"
        static let notificationsEnabled = "settings.notifications_enabled"
    }

    private let keychain: KeychainServicing

    @Published var privacyMode: Bool {
        didSet { writeBool(privacyMode, for: Keys.privacyMode) }
    }

    @Published var biometricEnabled: Bool {
        didSet { writeBool(biometricEnabled, for: Keys.biometricEnabled) }
    }

    @Published var soundEnabled: Bool {
        didSet { writeBool(soundEnabled, for: Keys.soundEnabled) }
    }

    @Published var messageTTL: MessageTTL {
        didSet { writeInt(messageTTL.rawValue, for: Keys.messageTTL) }
    }

    @Published var autoLockTimeout: AutoLockTimeout {
        didSet { writeInt(autoLockTimeout.rawValue, for: Keys.autoLock) }
    }

    /// User opt-in for receiving push notifications when contacts message
    /// while you're offline. Default is OFF — privacy-first, never silently
    /// register the device for APNs alerts. Setting this to `true` kicks off
    /// the iOS permission prompt and (on grant) `registerForRemoteNotifications`.
    @Published var notificationsEnabled: Bool {
        didSet { writeBool(notificationsEnabled, for: Keys.notificationsEnabled) }
    }

    init(keychain: KeychainServicing) {
        self.keychain = keychain
        self.privacyMode      = Self.readBool(keychain: keychain, key: Keys.privacyMode, default: false)
        self.biometricEnabled = Self.readBool(keychain: keychain, key: Keys.biometricEnabled, default: false)
        self.soundEnabled     = Self.readBool(keychain: keychain, key: Keys.soundEnabled, default: true)
        let ttlRaw = Self.readInt(keychain: keychain, key: Keys.messageTTL, default: MessageTTL.fiveMinutes.rawValue)
        self.messageTTL = MessageTTL(rawValue: ttlRaw) ?? .fiveMinutes
        let alRaw = Self.readInt(keychain: keychain, key: Keys.autoLock, default: AutoLockTimeout.oneMinute.rawValue)
        self.autoLockTimeout = AutoLockTimeout(rawValue: alRaw) ?? .oneMinute
        self.notificationsEnabled = Self.readBool(keychain: keychain, key: Keys.notificationsEnabled, default: false)
    }

    // MARK: - Private

    private func writeBool(_ value: Bool, for key: String) {
        try? keychain.set(Data([value ? 0x01 : 0x00]), for: key)
    }

    private func writeInt(_ value: Int, for key: String) {
        try? keychain.set(Data("\(value)".utf8), for: key)
    }

    private static func readBool(keychain: KeychainServicing, key: String, `default`: Bool) -> Bool {
        guard let data = try? keychain.get(key), data.count == 1 else { return `default` }
        return data[0] == 0x01
    }

    private static func readInt(keychain: KeychainServicing, key: String, `default`: Int) -> Int {
        guard let data = try? keychain.get(key),
              let str = String(data: data, encoding: .utf8),
              let n = Int(str) else { return `default` }
        return n
    }
}

File: android/app/src/main/java/com/kordar/ghostchat/core/managers/SettingsManager.kt
package com.kordar.ghostchat.core.managers

import com.kordar.ghostchat.core.security.KeystoreServicing
import com.kordar.ghostchat.models.AutoLockTimeout
import com.kordar.ghostchat.models.MessageTTL
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.nio.charset.StandardCharsets

/**
 * Keystore-backed user preferences. Matches iOS `SettingsManager` key names so preferences
 * on one platform remain semantically compatible if later migrated.
 */
class SettingsManager(private val keystore: KeystoreServicing) {

    private object Keys {
        const val PRIVACY_MODE          = "settings.privacy_mode"
        const val BIOMETRIC_ENABLED     = "settings.biometric_enabled"
        const val SOUND_ENABLED         = "settings.sound_enabled"
        const val MESSAGE_TTL           = "settings.message_ttl"
        const val AUTO_LOCK             = "settings.auto_lock"
        const val NOTIFICATIONS_ENABLED = "settings.notifications_enabled"
    }

    private val _privacyMode      = MutableStateFlow(readBool(Keys.PRIVACY_MODE, false))
    private val _biometricEnabled = MutableStateFlow(readBool(Keys.BIOMETRIC_ENABLED, false))
    private val _soundEnabled     = MutableStateFlow(readBool(Keys.SOUND_ENABLED, true))
    private val _messageTTL       = MutableStateFlow(
        MessageTTL.fromSeconds(readInt(Keys.MESSAGE_TTL, MessageTTL.FIVE_MINUTES.seconds))
    )
    private val _autoLockTimeout  = MutableStateFlow(
        AutoLockTimeout.fromSeconds(readInt(Keys.AUTO_LOCK, AutoLockTimeout.ONE_MINUTE.seconds))
    )
    private val _notificationsEnabled = MutableStateFlow(readBool(Keys.NOTIFICATIONS_ENABLED, false))

    val privacyMode: StateFlow<Boolean> = _privacyMode.asStateFlow()
    val biometricEnabled: StateFlow<Boolean> = _biometricEnabled.asStateFlow()
    val soundEnabled: StateFlow<Boolean> = _soundEnabled.asStateFlow()
    val messageTTL: StateFlow<MessageTTL> = _messageTTL.asStateFlow()
    val autoLockTimeout: StateFlow<AutoLockTimeout> = _autoLockTimeout.asStateFlow()

    /**
     * User opt-in for receiving push notifications when contacts message
     * while you're offline. Default OFF — privacy-first. Setting `true`
     * should be paired with the runtime POST_NOTIFICATIONS prompt on API 33+;
     * the UI layer (Settings toggle) handles that.
     */
    val notificationsEnabled: StateFlow<Boolean> = _notificationsEnabled.asStateFlow()

    fun setPrivacyMode(value: Boolean)      { _privacyMode.value = value;      writeBool(Keys.PRIVACY_MODE, value) }
    fun setBiometricEnabled(value: Boolean) { _biometricEnabled.value = value; writeBool(Keys.BIOMETRIC_ENABLED, value) }
    fun setSoundEnabled(value: Boolean)     { _soundEnabled.value = value;     writeBool(Keys.SOUND_ENABLED, value) }
    fun setMessageTTL(value: MessageTTL)    { _messageTTL.value = value;       writeInt(Keys.MESSAGE_TTL, value.seconds) }
    fun setAutoLockTimeout(v: AutoLockTimeout) { _autoLockTimeout.value = v;   writeInt(Keys.AUTO_LOCK, v.seconds) }
    fun setNotificationsEnabled(value: Boolean) {
        _notificationsEnabled.value = value
        writeBool(Keys.NOTIFICATIONS_ENABLED, value)
    }

    private fun readBool(key: String, default: Boolean): Boolean {
        val raw = keystore.get(key) ?: return default
        return raw.isNotEmpty() && raw[0] == 0x01.toByte()
    }
    private fun readInt(key: String, default: Int): Int {
        val raw = keystore.get(key) ?: return default
        return String(raw, StandardCharsets.UTF_8).toIntOrNull() ?: default
    }
    private fun writeBool(key: String, value: Boolean) =
        keystore.set(key, byteArrayOf(if (value) 0x01 else 0x00))
    private fun writeInt(key: String, value: Int) =
        keystore.set(key, value.toString().toByteArray(StandardCharsets.UTF_8))
}
